server.go

```diff
@@ -1,28 +1,168 @@
+
+// in case of TCP we only use defaultReadTimeout for the first read
+// then we start using defaultTCPIdleTimeout
+const defaultTCPIdleTimeout = 8 * time.Second
+
+// helper struct that is used in several SetReadDeadline calls
+var longTimeAgo = time.Unix(1, 0)
+
+// ServerDNSCrypt is an interface for a DNSCrypt server
+type ServerDNSCrypt interface {
+	// ServeTCP listens to TCP connections, queries are then processed by Server.Handler.
+	// It blocks the calling goroutine and to stop it you need to close the listener
+	// or call ServerDNSCrypt.Shutdown.
+	ServeTCP(l net.Listener) error
+
+	// ServeUDP listens to UDP connections, queries are then processed by Server.Handler.
+	// It blocks the calling goroutine and to stop it you need to close the listener
+	// or call ServerDNSCrypt.Shutdown.
+	ServeUDP(l *net.UDPConn) error
+
+	// Shutdown tries to gracefully shutdown the server. It waits until all
+	// connections are processed and only after that it leaves the method.
+	// If context deadline is specified, it will exit earlier
+	// or call ServerDNSCrypt.Shutdown.
+	Shutdown(ctx context.Context) error
+}
+
+// Server is a simple DNSCrypt server implementation
 type Server struct {
-	// ProviderName - DNSCrypt provider name
+	// ProviderName is a DNSCrypt provider name
 	ProviderName string
 
-	// ResolverCert - contains resolver certificate.
+	// ResolverCert contains resolver certificate.
 	ResolverCert *Cert
 
 	// Handler to invoke. If nil, uses DefaultHandler.
 	Handler Handler
+
+	// Shutdown handling
+	// --
+	lock         sync.RWMutex // protects access to all the fields below
+	started      bool
+	wg           sync.WaitGroup            // active workers (servers)
+	tcpListeners map[net.Listener]struct{} // track active TCP listeners
+	udpListeners map[*net.UDPConn]struct{} // track active UDP listeners
+	tcpConns     map[net.Conn]struct{}     // track active connections
 }
 
-// serveDNS - serves DNS response
-func (s *Server) serveDNS(rw ResponseWriter, r *dns.Msg) {
-	if r == nil || len(r.Question) != 1 || r.Response {
-		log.Tracef("Invalid query: %v", r)
+// type check
+var _ ServerDNSCrypt = &Server{}
+
+// prepareShutdown - prepares the server to shutdown:
+// unblocks reads from all connections related to this server
+// marks the server as stopped
+// if the server is not started, returns ErrServerNotStarted
+func (s *Server) prepareShutdown() error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if !s.started {
+		log.Info("Server is not started")
+		return ErrServerNotStarted
+	}
+
+	s.started = false
+
+	// Unblock reads for all active tcpConns
+	for conn := range s.tcpConns {
+		_ = conn.SetReadDeadline(longTimeAgo)
+	}
+
+	// These listeners were passed to us from the outside so we cannot close
+	// them here - this is up to the calling code to do that. Instead of that,
+	// we call Set(Read)Deadline to unblock goroutines that are currently
+	// blocked on reading from those listeners.
```

Comment:
I would place this comment above the loop where the `tcpConns` are processed.

Reply:
The same applies to listeners as well, why tcpConns only

Reply:
Exactly.  That is why I wanted the comment to be before all of them, and not in the middle of the process :-) .  That is:

```go
// Comment explaining the following actions.

action1()

action2()

action3()
```

As opposed to the current:


```go
action1()

// Comment explaining the following actions.

action2()

action3()
```